src/core/middlewares/countquota/handler.go

```diff
@@ -63,124 +42,86 @@ func New(next http.Handler) http.Handler {
+			return
+		}
+		mf.Exist = imageExist
+		if imageExist {
+			if af.Digest != mf.Digest {
+				mf.DigestChanged = true
 			}
-			req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
-
-			manifest, desc, err := distribution.UnmarshalManifest(mediaType, data)
-			if err != nil {
-				cqh.tryFreeTag()
-				log.Warningf("Error occurred when to Unmarshal Manifest %v", err)
-				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to Unmarshal Manifest %v", err)), http.StatusInternalServerError)
-				return
+		} else {
+			quotaRes := &quota.ResourceList{
+				quota.ResourceCount: 1,
 			}
-			cqh.mfInfo.Refrerence = manifest.References()
-			cqh.mfInfo.Digest = desc.Digest.String()
-
-			projectID, err := cqh.getProjectID(strings.Split(repository, "/")[0])
+			err := util.TryRequireQuota(mf.ProjectID, quotaRes)
 			if err != nil {
-				log.Warningf("Error occurred when to get project ID %v", err)
-				return
-			}
-			cqh.mfInfo.ProjectID = projectID
-
-			imageExist, af, err := cqh.imageExist()
-			if err != nil {
-				cqh.tryFreeTag()
-				log.Warningf("Error occurred when to check Manifest existence by repo and tag name %v", err)
-				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to check Manifest existence %v", err)), http.StatusInternalServerError)
-				return
-			}
-			cqh.mfInfo.Exist = imageExist
-			if imageExist {
-				if af.Digest != cqh.mfInfo.Digest {
-					cqh.mfInfo.DigestChanged = true
-				}
-			} else {
-				quotaRes := &quota.ResourceList{
-					quota.ResourceCount: 1,
-				}
-				err := cqh.tryRequireQuota(quotaRes)
-				if err != nil {
-					cqh.tryFreeTag()
-					log.Errorf("Cannot get quota for the manifest %v", err)
-					if err == ErrRequireQuota {
-						http.Error(rw, util.MarshalError("StatusNotAcceptable", fmt.Sprintf("Cannot get quota for the manifest %v", err)), http.StatusNotAcceptable)
-						return
-					}
-					http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to require quota for the manifest %v", err)), http.StatusInternalServerError)
+				cqh.tryFreeTag(mf)
+				log.Errorf("Cannot get quota for the manifest %v", err)
+				if err == util.ErrRequireQuota {
+					http.Error(rw, util.MarshalError("StatusNotAcceptable", "Your request is reject as not enough quota."), http.StatusNotAcceptable)
 					return
 				}
-				cqh.mfInfo.Quota = quotaRes
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to require quota for the manifest %v", err)), http.StatusInternalServerError)
+				return
 			}
-
-			*req = *(req.WithContext(context.WithValue(req.Context(), util.MFInfokKey, mfInfo)))
+			mf.Quota = quotaRes
 		}
-
+		*req = *(req.WithContext(context.WithValue(req.Context(), util.MFInfokKey, mfInfo)))
 	}
 
 	cqh.next.ServeHTTP(rw, req)
```

Comment:
I think that we rollback the quota in  the same handler is better when the response code != 200?

Reply:
add the handleResponse for size and count